designs/2025-css-vars-tracking/README.md

```diff
@@ -0,0 +1,128 @@
+- Repo: eslint/css
+- Start Date: 2025-06-24
+- RFC PR:
+- Authors: Nicholas C. Zakas
+
+# CSS Custom Property Tracking in SourceCode
+
+## Summary
+
+This RFC proposes to add capabilities to `CSSSourceCode` to track CSS custom properties (also known as CSS variables). This includes identifying where custom properties are defined and where they are used, as well as providing a way to retrieve the value of a custom property at a specific location in the code.
+
+## Motivation
+
+CSS custom properties are a foundational part of modern CSS development. They allow for more modular and maintainable stylesheets. For a CSS linter to provide accurate and helpful rules, it needs to be able to understand how custom properties are being used. Currently, any rule that needs to validate property values must implement its own logic for tracking custom properties, leading to duplicated effort and potential inconsistencies.
+
+By building this functionality directly into `CSSSourceCode`, we can provide a consistent and reliable way for all rules to access information about custom properties. Uses of this information include:
+
+*   `no-invalid-properties` rule
+*   `font-family-fallbacks` rule
+*   Detecting references to undefined custom properties.
+*   Future: Detecting unused custom properties.
+
+This change is based on the discussion in [eslint/css#160](https://github.com/eslint/css/issues/160).
+
+**Note:** The logic described in this RFC is already implemented in the [`no-invalid-properties`](https://github.com/eslint/css/blob/main/docs/rules/no-invalid-properties.md) rule. This proposal wants to standardize the logic and make it available to all rules.
+
+## Detailed Design
+
+The proposed changes are implemented in the [`CSSSourceCode`](https://github.com/eslint/css/blob/main/src/languages/css-source-code.js) class.
+
+### `customProperties` Map
+
+```ts
+interface CustomPropertyUses {
+    declarations: Array<DeclarationPlain>;
+    definitions: Array<AtrulePlain>;
+    references: Array<FunctionNode>;
+}
+
+interface CSSSourceCode {
+    #customProperties: Map<string, CustomPropertyUses>
+}
+```
+
+A new public property, `#customProperties`, will be added to `CSSSourceCode`. This will be a `Map` where the keys are the custom property names (e.g., `--my-color`) and the values are `CustomPropertyUses` objects. The `CustomPropertyUses` class will have two properties:
```

Comment:
```suggestion
A new public property, `#customProperties`, will be added to `CSSSourceCode`. This will be a `Map` where the keys are the custom property names (e.g., `--my-color`) and the values are `CustomPropertyUses` objects. The `CustomPropertyUses` class will have three properties:
```